### ISSUE TITLE
Mismatching virtual package error should hint to solution (set `system-requirements`)

### ISSUE THREAD

**Issue by pavelzw** [prefix-dev/pixi#346]
### Problem description

```
$ pixi init
$ pixi add python
$ pixi install
x The current system has a mismatching virtual package. The project requires '__linux' to be at least version '5.10' but the system has version '4.18.0'
```

should be something like

```
$ pixi init
$ pixi add python
$ pixi install
x The current system has a mismatching virtual package. The project requires '__linux' to be at least version '5.10' but the system has version '4.18.0'. Try setting the following in your pixi.toml:
[system-requirements]
linux="4.18"
```

Similar for other virtual packages.

**Comment by jpursell** [comment]
This should be re-opened. I just got this unhelpful error

```
bash-4.4$ pixi shell
  × The current platform does not satisfy the minimal virtual package requirements
  ╰─▶ the project does not support 'linux-64'
```
pixi version 0.20.1

**Comment by pavelzw** [comment]
@ruben-arts can you reopen please? I don't have permissions to do so

**Comment by jpursell** [comment]
I should mention that the fix for this was not to add something to [system-requirements], but rather I needed to add "linux-64" to the platforms. It be very helpful if pixi would tell the user that.

**Comment by roaldarbol** [comment]
I also just got this for a `pypi` install:
```py
❯ pixi add --pypi syncstart==1.1.1
  × failed to solve the pypi requirements of 'default' 'osx-64'
  ├─▶ Failed to build: `opencv-python==4.9.0.80`
  ├─▶ Failed to install requirements from build-system.requires (install)
  ├─▶ Failed to download and build distributions
  ├─▶ Failed to fetch wheel: numpy==1.22.2
  ├─▶ Failed to build: `numpy==1.22.2`
  ╰─▶ Build backend failed to build wheel through `build_wheel()` with exit status: 1
  ```
... along with the longer `stdout` stuff.
It was resolved by setting:
```yaml
[system-requirements]
mac0s = "11.0"

**Comment by ruben-arts** [comment]
Honestly the users should feel the power of system requirements and not the downside. I've got a proposal to solve this issue with some workflow improvements:

TL;DR: Make pixi smarter and let it interact with the dependency resolution automatically, and help the user by doing the work for them.

### Definitions
**Virtual packages:** a dependency of a package that can not be provided as a conda package, these are system specifications or libc implementation on the system. e.g. `cuda`, `linux`, `macos`, `glibc` etc.
**System requirements**: Pixi's take on defining a virtual package in the project's manifest. It defines the requirements the project has of the machine it needs to run on, which it can't provide itself. There is a set of defaults per platform, these are used during the solve of the dependencies to avoid the solve to only work on your specific machine. If the project requires a different set then the default they need to be defined in the `[system-requirements]` table. 

### Solution

Let's make pixi solve with the current available virtual packages. Then after the solve figure out what virtual packages are required. If those differ from the default virtual packages they will automatically be added to the manifest file as system requirements. This goes both ways, having higher or lower versions on your current machine.

### Solution to newly created problem

The down side of change would be that the solution(lock file) might only work for the latest machines as pixi picks the latest versions available for the given virtual packages. Potentially lowering the amount of machines your project can run on. This could be fixed by giving pixi an interactive option to re-solve for your machine, skipping the system-requirements given by the project. When it finds a different solution for your machine it automatically writes it back to the manifest, allowing the users to track these changes in vcs. 

Example of the interactive prompt:
```
> pixi install
The project requires your system to have `cuda 12.0`. 
Should I try to solve a new environment with this machine's specs?
 > Yes
 - No
```

Change I expect to be needed:
- Pixi needs to share the machines virtual packages instead of the projects requirements with the solver.
- Rattler needs to share the virtual packages of the required packages back to pixi. (probably an API break :( )
- The cli needs to get an option to bypass the interactive request, e.g. `--ignore-system-requirements`

Questions I don't have an answer to yet:
- Should pixi try it's first solve with the default set of virtual packages to avoid changing the default set if not required?
- Would the same thing be possible for the `pypi-dependencies`?


ping @baszalmstra @wolfv for input

**Comment by roaldarbol** [comment]
I think it IS really powerful, so I hope it will become super intuitive! For lots of tracking software (e.g. https://github.com/DeepLabCut/DeepLabCut, https://gitlab.com/polavieja_lab/idtrackerai) *installation* has been a major sticking point for adoption, especially since they depend on TensorFlow and thus getting `cuda` just right across platforms - this would be greatly improved by a simple specification.

For me personally, I think one of the confusing aspects have been around the OS-named `system-requirements`, and how they differ from `platforms`. E.g. if I have already specified `osx-64` as the platform, it's been unintuitive why I'd need to additionally specify `macos = "11.0"` under `system-requirements`. Usually, the version requirements is to narrow the solving, so it took me by surprise that you could get something to solve by specifying a version - I'd think it would otherwise just attempt to solve against all system requirements for the specified platform. 

Lastly, is there no way it's possible for pixi to "figure out" the correct requirement for the user's current platform? E.g. for `cuda`, It would make life easier for *users* if they didn't have to know the CUDA requirements for their particular platform. Here the target audience are *not* super tech-savvy, but would like to use techy tools easily.

**Comment by baszalmstra** [comment]
I think your idea can work @ruben-arts but I cant really see the ramifications yet. Lets try to build an experiment in the near future.

**Comment by ruben-arts** [comment]
Hey @roaldarbol, Thanks for the input, this is good to take into account in the UX design. I believe that the things you mention are all atleast what I tried to explain in my proposal. So lets see what we can come up with in the first version!

**Comment by ruben-arts** [comment]
@zachcp Yep exact same issue. 

Try:

```toml
# [tool.pixi.system-requirements] for pyproject.toml
[system-requirements]
linux = "5.4"
```

**Comment by zoobab** [comment]
I have this issue on a centos 7 distro with an old kernel Linux version 3.10.0-1160.123.1.el7.x86_64:

```
$ pixi run start
  × The current system has a mismatching virtual package. The project requires '__linux' to be at least
  │ version '4.18' but the system has version '3.10.0'
```

**Comment by noamgot** [comment]
Encountered this one today (in a gitlab-ci pipeline):
```
$ pixi run py-check
Pixi task (py-check in lint): ruff check; ruff format --check
  × The current system has a mismatching virtual package. The project requires
  │ '__linux' to be at least version '4.18' but the system has version
  │ '4.15.0'
```

It failed despite setting in `pixi.toml`:
```toml
[system-requirements]
linux = "4.15"
```

Since my project also contains `pyproject.toml` file, I also tried to remove the former change and write in `pyprpoject.toml`:
```toml
[tool.pixi.system-requirements]
linux = "4.15"
```

both failed with the error mentioned above.

**Comment by vsoch** [comment]
I just ran into this for cuda:

```console
 > [10/10] RUN . /root/.bash_profile &&     pixi run --manifest-path ./pixi.toml ./build.sh:                                        
0.180 
0.180   × The platform you are running on should at least have the virtual package
0.180   │ __cuda on version 12.2.0, build_string: 0
0.180 
```
I'm not sure how to fix it - the base container I'm using is that exact version of cuda.

**Comment by vsoch** [comment]
I'm trying to install this gromacs version with cuda:

![Image](https://github.com/user-attachments/assets/b52cd5e4-4448-42d7-b4ea-b28dae78e9f0)

There aren't any instructions for how to target a specific build via the pixi.toml. I'm trying downloading the .conda and seeing if pixi will install that directly.

**Comment by diallobakary4** [comment]
I do have this in .toml dependencies section for gromacs

`gromacs = { version = "==2022.4", build = "nompi_hca75aac_100" , channel="conda-forge" }
`

**Comment by hunger** [comment]
#6178 fixes this. It clearly lists the missing virtual packages and such.

**Comment by hunger** [comment]
#6178 has landed, let's close this.
